drivers/flash/nrf_qspi_nor.c

```diff
@@ -503,27 +505,95 @@ static inline int qspi_nor_read_id(struct device *dev,
 	return 0;
 }
 
+static inline nrfx_err_t read_non_aligned(struct device *dev, off_t addr,
+					  void *dest, size_t size)
+{
+	uint8_t __aligned(4) buf[WORD_SIZE];
+	uint8_t *dptr = dest;
+
+	off_t flash_prefix = WORD_SIZE - (addr % WORD_SIZE);
+
+	if (flash_prefix > size) {
+		flash_prefix = size;
+	}
+
+	off_t dest_prefix = WORD_SIZE - (off_t)dptr % WORD_SIZE;
+
+	if (dest_prefix > size) {
+		dest_prefix = size;
+	}
+
+	off_t flash_suffix = (size - flash_prefix) % WORD_SIZE;
+	off_t flash_middle = size - flash_prefix - flash_suffix;
+
+	off_t shift = flash_prefix - dest_prefix;
+	nrfx_err_t res = NRFX_SUCCESS;
+
+	/* read from aligned flash to aligned memory */
+	if (flash_middle != 0) {
+		res = nrfx_qspi_read(dptr + dest_prefix, flash_middle,
+				     addr + flash_prefix);
+		qspi_wait_for_completion(dev, res);
+		if (res != NRFX_SUCCESS) {
+			return res;
+		}
+	}
+
+	/* perform shift in RAM */
+	if (shift > 0) {
+		/* destination is at higher address than source */
+		uint8_t *shift_src = dptr + dest_prefix + flash_middle - 1;
+		uint8_t *shift_dest = dptr + dest_prefix +
+				      flash_middle + shift - 1;
+
```

Comment:
If you would use prefix '--' in while loop, you could get rid of the '-1' in both calculations and reduce the condition to 'shift_dest != dptr + dest_prefix'.
This has also benefit because postfix `--` and '++' require saving value of the operand betwen sequence points while prefix variants do not require such operation.

Reply:
And place `dptr + dest_prefix` in parentheses.

Reply:
Btw, try to predict result of below code and then run it under different compilers (gcc, clang, tcc, etc):
```
#include <stdio.h>

int main() {
    int i = 0;
    printf("%d %d %d %d\n", i++, i++, i++, i++);
    i = 0;
    printf("%d %d %d %d\n", ++i, ++i, ++i, ++i);
}
```
That would be tricky recruitment question ;)

Reply:
Yeah, it is nice to know how used tools work :))